Add smoothly broken power law to adjust worldtube radius

nikwit opened the pull request.

PunJustice commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/Tags.hpp:
```diff
@@ -102,6 +102,72 @@ struct SelfForceOptions {
+  struct Exponent {
+    using type = double;
+    static constexpr Options::String help{
+        "The exponent according to which the excision sphere grows with "
```
Can you specify this is \alpha like you do for the TransitionWidth.

PunJustice commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/Tags.hpp:
```diff
@@ -102,6 +102,72 @@ struct SelfForceOptions {
+    using type = double;
+    static constexpr Options::String help{
+        "The transition radius rb of the smoothly broken power law. At this "
+        "point the transitions to a constant value."};
```
the radius transitions to a constant value.

PunJustice commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/Tags.hpp:
```diff
@@ -102,6 +102,72 @@ struct SelfForceOptions {
+  struct Amplitude {
+    using type = double;
+    static constexpr Options::String help{
+        "The amplitude of the smoothly broken power law."};
```
The amplitude A

PunJustice commented on tests/Unit/Evolution/Systems/CurvedScalarWave/Worldtube/Test_RadiusFunctions.cpp:
```diff
@@ -0,0 +1,78 @@
+#include "Framework/TestHelpers.hpp"
+
+namespace CurvedScalarWave::Worldtube {
+namespace {
```
If you wanted one more test you could reimplement these in python and test against them, but I think these tests are already quite good.

nikwit replied:
yes, the python tests are usually a good idea if there is no other test available but I think the ones there cover all regions quite well.

PunJustice commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/RadiusFunctions.hpp:
```diff
@@ -0,0 +1,40 @@
+namespace CurvedScalarWave::Worldtube {
+
+/*!
+ * \brief A smoothly broken power law, that grows falls off to a constant value
```
that grows->thats growth, or reword.

PunJustice commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/RadiusFunctions.cpp:
```diff
@@ -0,0 +1,57 @@
+                               double rb, double delta) {
+  if (alpha < 0. or alpha > 4.) {
+    ERROR(
+        "Only exponents between 0 and 4 have been tested for the broken law. "
```
broken power law

nikwit commented:
thanks, I incorporated your changes and changed the order of the parameters to something more natural. 

I also increased the minimum value of delta from 1e-3 to 1e-2. This seemed to cause an FPE in the CI for all compilers which I was not able to reproduce with clang or gcc locally or on a cluster. Probably something hardware specific.

knelli2 commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/RadiusFunctions.hpp:
```diff
@@ -0,0 +1,40 @@
+
+#pragma once
+
+namespace CurvedScalarWave::Worldtube {
+
```
Remove extra line breaks

knelli2 commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/RadiusFunctions.cpp:
```diff
@@ -0,0 +1,56 @@
+#include "Utilities/ErrorHandling/Error.hpp"
+
+namespace CurvedScalarWave::Worldtube {
+
+double smooth_broken_power_law(double orbit_radius, double alpha, double amp,
+                               double rb, double delta) {
```
[optional] Make these all `const` since you don't change them. Also can remove the extra space above

knelli2 commented on src/Evolution/Systems/CurvedScalarWave/Worldtube/RadiusFunctions.cpp:
```diff
@@ -0,0 +1,56 @@
+
+double smooth_broken_power_law(double orbit_radius, double alpha, double amp,
+                               double rb, double delta) {
+  if (alpha < 0. or alpha > 4.) {
+    ERROR(
```
Factor out these two checks into a `detail` namespace so they share the same wording for both this function and the deriv

knelli2 commented on tests/Unit/Evolution/Systems/CurvedScalarWave/Worldtube/Test_RadiusFunctions.cpp:
```diff
@@ -0,0 +1,78 @@
+
+void test_smooth_broken_power_law_derivative() {
+  const double amp = 1.234;
+  const double rb = 10.;
+  const Approx custom_approx = Approx::custom().epsilon(1.e-10).scale(1.0);
```
Is this larger approx because of the wide range of radii tested below? Whatever the reason, maybe add a small comment explaining.

nikwit replied:
The comparison is done with a finite difference method which I found to be accurate to about 1e-10. I added a small comment.

knelli2 reviewed (changes requested):
Squash directly along with the current fixup for @PunJustice's changes

nikwit committed:
Add excision radii tags

nikwit committed:
Add broken power law radius function

nikwit commented:
Sorry for the delay. I squashed your suggestions in, thank you for your reviews!
